Spigot-Server-Patches/0619-Add-per-player-block-reach.patch

```diff
@@ -0,0 +1,129 @@
+                 // Paper - move check up
+                 this.player.clearActiveItem(); // SPIGOT-4706
+diff --git a/src/main/java/net/minecraft/server/PlayerInteractManager.java b/src/main/java/net/minecraft/server/PlayerInteractManager.java
+index 485b609bb5387b0f8a46c1201177cdc6d183ad91..2bec046d2d618348718e10d13d1d4ceef08b6cba 100644
+--- a/src/main/java/net/minecraft/server/PlayerInteractManager.java
++++ b/src/main/java/net/minecraft/server/PlayerInteractManager.java
+@@ -129,7 +129,7 @@ public class PlayerInteractManager {
+         double d2 = this.player.locZ() - ((double) blockposition.getZ() + 0.5D);
+         double d3 = d0 * d0 + d1 * d1 + d2 * d2;
+ 
+-        if (d3 > 36.0D) {
++        if (d3 > this.player.getBukkitEntity().getMaxBlockDestroyReachSquared()) { // Paper - use individual player block reach
+             this.player.playerConnection.sendPacket(new PacketPlayOutBlockBreak(blockposition, this.world.getType(blockposition), packetplayinblockdig_enumplayerdigtype, false, "too far"));
+         } else if (blockposition.getY() >= i) {
+             this.player.playerConnection.sendPacket(new PacketPlayOutBlockBreak(blockposition, this.world.getType(blockposition), packetplayinblockdig_enumplayerdigtype, false, "too high"));
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+index 58caa3240b90cdc661e1e32e3f5c312ed62c3c21..d49a22c9e5b9764a7c319197800ad37f34f7a90c 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+@@ -146,6 +146,12 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+     private String resourcePackHash;
+     private static final boolean DISABLE_CHANNEL_LIMIT = System.getProperty("paper.disableChannelLimit") != null; // Paper - add a flag to disable the channel limit
+     private long lastSaveTime;
++    private int survivalBlockReach = 6;
++    private int survivalBlockReachSquared = 36;
++    private int creativeBlockReach = 7;
++    private int creativeBlockReachSquared = 49;
++    private double maxBlockPlaceReachSquared = 64D;
++    private double maxBlockDestroyReachSquared = 36D;
+     // Paper end
+ 
+     public CraftPlayer(CraftServer server, EntityPlayer entity) {
+@@ -2100,6 +2106,59 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+             ? (org.bukkit.entity.Firework) entity.getBukkitEntity()
+             : null;
+     }
++
++    @Override
++    public int getSurvivalBlockRange() {
++        return survivalBlockReach;
++    }
++
++    public int getSurvivalBlockReachSquared() {
++        return survivalBlockReachSquared;
++    }
++
++    @Override
++    public void setSurvivalBlockRange(int range) {
++        range = Math.max(6, range); // Ensure range can't be set lower than vanilla defaults, or else block ghosting will occur for vanilla clients
```

Comment:
After some discussion with someone else - should this be replaced with a `Preconditions` check?

Reply:
Yes. Existing API design prefers throwing and letting plugins deal with the consequences over silently adjusting the value.
Though like everything, there are exceptions.

Reply:
Addressed.